nomos-services/data-availability/network/src/backends/libp2p/common.rs

```diff
@@ -179,3 +181,67 @@ pub(crate) async fn handle_balancer_command<Stats: Debug>(
         error!("Error stats request: {cmd:?}");
     }
 }
+
+use arc_swap::ArcSwap;
+
+pub struct SwappableMembershipHandler<T: MembershipHandler> {
+    inner: ArcSwap<T>,
+}
+
+impl<T: MembershipHandler> SwappableMembershipHandler<T> {
+    pub fn new(handler: T) -> Self {
+        Self {
+            inner: ArcSwap::new(Arc::new(handler)),
+        }
+    }
+
+    pub fn update(&self, new_handler: T) {
+        self.inner.store(Arc::new(new_handler));
+    }
+
+    pub fn inner(&self) -> Arc<T> {
+        self.inner.load_full()
+    }
+}
+
+// Implement MembershipHandler for SwappableMembershipHandler
+impl<T: MembershipHandler> MembershipHandler for SwappableMembershipHandler<T> {
+    type NetworkId = T::NetworkId;
+    type Id = T::Id;
+
+    fn membership(&self, id: &Self::Id) -> HashSet<Self::NetworkId> {
+        self.inner.load().membership(id)
+    }
+
+    fn is_allowed(&self, id: &Self::Id) -> bool {
+        self.inner.load().is_allowed(id)
+    }
+
+    fn members_of(&self, network_id: &Self::NetworkId) -> HashSet<Self::Id> {
+        self.inner.load().members_of(network_id)
+    }
+
+    fn members(&self) -> HashSet<Self::Id> {
+        self.inner.load().members()
+    }
+
+    fn last_subnetwork_id(&self) -> Self::NetworkId {
+        self.inner.load().last_subnetwork_id()
+    }
+
+    fn get_address(&self, peer_id: &PeerId) -> Option<Multiaddr> {
+        self.inner.load().get_address(peer_id)
+    }
+
+    fn rebuild_with(
+        &self,
+        members: Vec<PeerId>,
+        addressbook: std::collections::HashMap<PeerId, Multiaddr>,
+    ) -> Self {
+        let inner = self.inner.load_full();
+        let inner = inner.rebuild_with(members, addressbook);
+        Self {
+            inner: ArcSwap::new(inner),
+        }
+    }
+}
```

Comment:
why this? Behaviours gonna share it? Who takes responsability of the swap then?

Reply:
Behaviors share the read only `MembershipHandler`, but da network service sees `SwappableMembershipHandler` and can use update to update it (swap the whole membership with the new one, and behaviors will just see the new one since then).